### ISSUE TITLE
[CMake] _TRACE is not defined in a Debug build

### ISSUE THREAD

**Issue by delcypher** [Z3Prover/z3#522]
@NikolajBjorner mentioned this issue to me recently. The Python build system defines ``_TRACE`` in a Debug build but the CMake build system does not.

This was a deliberate decision on my part and I want to explain the rationale:

The CMake build system provides the option ``ENABLE_TRACING`` which is off by default. When this option is enabled ``_TRACE`` is added as a compiler definition. This option is currently independent of the build type (e.g. ``Release``, ``Debug``).

It is not always possible to know the build configuration (e.g. ``Debug``, ``Release``) when CMake is run due to multi-configuration generators (e.g. Visual Studio). This means it is not possible force ``ENABLE_TRACING`` to ``ON`` when doing a debug build. However it is possible (using something CMake calls "generator expressions") to have the ``_TRACE`` define to be added depending to build configuration. The code for this already there but it is commented out

```cmake
# Should we always enable tracing when doing a debug build?
#list(APPEND Z3_COMPONENT_CXX_DEFINES $<$<CONFIG:Debug>:_TRACE>)
```

If uncommented this code it would lead to confusing behaviour because tracing would be enabled in a debug build even if ``ENABLE_TRACING`` is set to ``OFF``. This is why the code is currently commented out.

I can fix this but first I need to know

@NikolajBjorner @wintersteiger 
Is it ever desirable to have a debug build without tracing?

If the answer is yes then I think I should leave the CMake build system as it is to allow building a debug build without tracing. If the answer is no then I think I should rename the CMake option to ``ENABLE_TRACING_IN_NON_DEBUG_BUILD``. Then I would then set the build system to always define ``_TRACE`` in a debug build. The ``ENABLE_TRACING_IN_NON_DEBUG_BUILD`` would no longer be confusing because it would only matter in non debug builds.

**Comment by NikolajBjorner** [comment]
Thanks, for the explanations. Seems all solvable using a suitable script to set up the make-files.

> Is it ever desirable to have a debug build without tracing?
No: the trace macro is very cheap compared to debug assert checking. We want it all.

**Comment by NikolajBjorner** [comment]
Somehow, I still don't get tracing when using:

      cmake ENABLE_TRACING=TRUE -G "Ninja" ../

**Comment by agurfinkel** [comment]
cmake -DENABLE_TRACING=TRUE -G "Ninja" ../

**Comment by delcypher** [comment]
@NikolajBjorner 
>> Is it ever desirable to have a debug build without tracing?

> No: the trace macro is very cheap compared to debug assert checking. We want it all.

Thanks for clarifying. I'll implement a fix.

**Comment by delcypher** [comment]
@NikolajBjorner 

Just a useful hint. If you want to change an option in an existing build directory run

```
make edit_cache
```

(replace ``make`` with what ever generator you decided to use e.g.``ninja`` or ``nmake``)

This will invoke ``cmake-gui``, ``ccmake`` or ``cmake`` in an interactive mode (depending on what tools you have installed. These tools provide a user interface tweaking options. Once you have changed an option you will need to press the ``configure`` button and then ``generate`` button so that CMake regenerates the build files.

@agurfinkel Is right though. You need to use ``-DENABLE_TRACING=TRUE``, not ``ENABLE_TRACING=TRUE`` on the command line.

**Comment by wintersteiger** [comment]
Closing this as fixed in #525.
